doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,365 @@
+- Decides on version up front, reducing low level trait by trait branching.
+
+### Cons
+- Complex dance needs to be done up front, people may end up just
+  skipping it.
+- Still muddies the water for naive call store and recall managers. Does the
+  declaration of a supported version set implicitly exclude all other
+  traits in other traits & specs libraries. Reintroduces the spectre of optionality.
+- Manager reporting a supported schema version could imply every trait
+in that schema is supported, despite the need to still call
+managementPolicy.
+
+
+### Option 2 - In data
+
+In this conception, version information is sent "over the wire" so to
+speak. Each trait has a version packed into it, which is unpacked on
+the other side, and then used for any version branching.
+
+> **Note**
+>
+> Practically, this would likely be packed into the `traitId` of each
+> trait, something like
+> "`openassetio-mediacreation-v-4:LocatableContent`". This would
+> effectively make each version of a trait a new trait as far as the
+> `openassetio` core library is concerned.
+
+
+### Host side
+```python
+# The host side implementation is almost entirely unchanged from
+# from standard, excepting a new BatchElementError type
+trait_set = {LocatableContentTrait.kId}
+context = self.createTestContext()
+
+result = [None]
+
+def success_cb(idx, traits_data):
+    result[0] = traits_data
+
+def error_cb(idx, batchElementError):
+    if instanceOf(batchElementError, VersionNotSupportedError):
+      incompatible_with_manager_error()
+    ...
+
+self._manager.resolve(
+    [entity_reference], trait_set, ResolveAccess.kRead, context, success_cb, error_cb
+)
+```
+### Manager side
+```python
+# On the manager side, each trait gets unpacked (probably via a utility
+# method provided by traitgen or openassetio), and branching is done
+# on a trait by trait basis
+
+def resolve(
+        self,
+        entityReferences,
+        traitSet,
+        resolveAccess,
+        context,
+        hostSession,
+        successCallback,
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- Both managers and hosts have the option to be backwards compatible.
```

Comment:
[consider] I think I'd remove this, since negotiation also has this Pro, it's just that backward compatibility is ahead-of-time vs. just-in-time?

Reply:
Yeah you're right, holdover from when there was more than 2 options. deae53fb12c77c2f16365f60114d7a2d92697101